packages/flutter/lib/src/widgets/pop_scope.dart

```diff
@@ -0,0 +1,137 @@
+///
+/// If [canPop] is false, then a system back gesture will not pop the route off
+/// of the enclosing [Navigator]. [onPopInvoked] will still be called, and
+/// `didPop` will be `false`.
+///
+/// If [canPop] is true, then a system back gesture will cause the enclosing
+/// [Navigator] to receive a pop as usual. [onPopInvoked] will be called with
+/// `didPop` as `true`, unless the pop failed for reasons unrelated to
+/// [PopScope], in which case it will be `false`.
+///
+/// {@tool dartpad}
+/// This sample demonstrates how to use this widget to handle nested navigation
+/// in a bottom navigation bar.
+///
+/// ** See code in examples/api/lib/widgets/pop_scope/pop_scope.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [NavigatorPopHandler], which is a less verbose way to handle system back
+///    gestures in simple cases of nested [Navigator]s.
+///  * [Form.canPop] and [Form.onPopInvoked], which can be used to handle system
+///    back gestures in the case of a form with unsaved data.
+///  * [ModalRoute.registerPopEntry] and [ModalRoute.unregisterPopEntry],
+///    which this widget uses to integrate with Flutter's navigation system.
+class PopScope extends StatefulWidget {
+  /// Creates a widget that registers a callback to veto attempts by the user to
+  /// dismiss the enclosing [ModalRoute].
+  const PopScope({
+    super.key,
+    required this.child,
+    this.canPop = true,
+    this.onPopInvoked,
+  });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// {@template flutter.widgets.PopScope.onPopInvoked}
+  /// Called after a route pop was handled.
+  /// {@endtemplate}
+  ///
+  /// It's not possible to prevent the pop from happening at the time that this
+  /// method is called; the pop has already happened. Use [canPop] to
+  /// disable pops in advance.
+  ///
+  /// This will still be called even when the pop is canceled. A pop is canceled
+  /// when the relevant [Route.popDisposition] returns false, such as when
+  /// [canPop] is set to false on a [PopScope]. The `didPop` parameter
+  /// indicates whether or not the back navigation actually happened
+  /// successfully.
+  ///
+  /// See also:
+  ///
+  ///  * [Route.onPopInvoked], which is similar.
+  final PopInvokedCallback? onPopInvoked;
+
+  /// {@template flutter.widgets.PopScope.canPop}
+  /// When false, blocks the current route from being popped.
+  ///
+  /// This includes the root route, where upon popping, the Flutter app would
+  /// exit.
+  ///
+  /// If multiple [PopScope] widgets appear in a route's widget subtree, then
+  /// each and every `canPop` must be `true` in order for the route to be
+  /// able to pop.
+  ///
+  /// [Android's predictive back](https://developer.android.com/guide/navigation/predictive-back-gesture)
+  /// feature will not animate when this boolean is false.
+  /// {@endtemplate}
+  final bool canPop;
```

Comment:
With the old `WillPopScope`, the API was expecting a Future<bool>, which enabled hooking this into showing a dialog before popping the route. How would we support that here?

Reply:
I forgot to respond here because we discussed this on Discord, but for the record, I've added an example to the migration guide about this: https://github.com/flutter/website/pull/9244.  It's similar to the pop_scope.0.dart example in this PR.

Reply:
@justinmc it work in case pop to previous route. But can you please example how to hide the app when double-back at first route?

Before
```
  Future<bool> onWillPop() {
    DateTime now = DateTime.now();
    if (_currentBackPressTime == null ||
        now.difference(_currentBackPressTime!) > const Duration(seconds: 2)) {
      _currentBackPressTime = now;
      DialogUtil.showSnackBar(context, content: 'Press back again to exit');
      return Future.value(false);
    }
    return Future.value(true);
  }
```

I can not do it with new PopScope, the only way required setState to change `canPop`